cpp/src/arrow/util/decimal.cc

```diff
@@ -92,6 +93,12 @@ struct DecimalRealConversion : public BaseDecimalRealConversion {
     constexpr int kMantissaBits = RealTraits<Real>::kMantissaBits;
     constexpr int kMantissaDigits = RealTraits<Real>::kMantissaDigits;
 
+    // to avoid precision and rounding issues, we'll unconditionally
+    // throw Decimal32 to the approx algorithm instead.
```

Comment:
Can we open a GH issue to improve this and reference it here?
(for example, a very simple solution would be to convert to Decimal64 and convert down to Decimal32 afterwards)

Reply:
Issue opened and referenced here (#44216)